src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs

```diff
@@ -307,26 +307,40 @@ private static string GetRestoreProjectPath(IVsTargetFrameworks tfms)
             return GetNonEvaluatedPropertyOrNull(tfms, RestorePackagesPath, e => e);
         }
 
+        /// <summary>
+        /// The result will contain CLEAR and no sources specified in RestoreSources if the clear keyword is in it.
+        /// If there are additional sources specified, the value RestoreAdditionalProjectSources will be set in the result and then all the additional sources will follow
+        /// </summary>
         private static string[] GetRestoreSources(IVsTargetFrameworks tfms)
         {
             var sources = MSBuildStringUtility.Split(GetNonEvaluatedPropertyOrNull(tfms, RestoreSources, e => e));
 
             sources = HandleClear(sources);
 
             var additional = MSBuildStringUtility.Split(GetNonEvaluatedPropertyOrNull(tfms, RestoreAdditionalProjectSources, e => e));
-            sources = sources.Concat(additional).ToArray();
-
+            if (additional.Length != 0)
+            {
+                sources = (sources.Concat(new string[] { RestoreAdditionalProjectSources }).Concat(additional)).ToArray();
+            }
             return sources;
         }
 
+        /// <summary>
+        /// The result will contain CLEAR and no sources specified in RestoreFallbackFolders if the clear keyword is in it.
+        /// If there are additional fallback folders specified, the value RestoreAdditionalProjectFallbackFolders will be set in the result and then all the additional fallback folders will follow
+        /// </summary>
         private static IEnumerable<string> GetRestoreFallbackFolders(IVsTargetFrameworks tfms)
         {
             var folders = MSBuildStringUtility.Split(GetNonEvaluatedPropertyOrNull(tfms, RestoreFallbackFolders, e => e));
 
             folders = HandleClear(folders);
 
             var additional = MSBuildStringUtility.Split(GetNonEvaluatedPropertyOrNull(tfms, RestoreAdditionalProjectFallbackFolders, e => e));
-            return folders.Concat(additional);
+            if (additional.Length != 0)
+            {
+                folders = (folders.Concat(new string[] { RestoreAdditionalProjectFallbackFolders }).Concat(additional)).ToArray();
```

Comment:
A helper method for this in the same class as the reader would be less error prone